### ISSUE TITLE
Error: S3 client error. uninitialized constant Aws::S3

### ISSUE THREAD

**Issue by lancelakey** [packetloop/fluent-plugin-cloudfront-log-v0.14-fix#5]
I'm running the latest version of Fluentd in a Docker container with the following plugins added:
fluent-plugin-cloudfront-log-v0.14-fix
fluent-plugin-elasticsearch

When I launch fluentd I get the following error.
Any ideas about a likely cause and fix?

# Error: S3 client error. uninitialized constant Aws::S3
```
2018-10-05 23:42:56 +0000 [info]: #0 starting fluentd worker pid=29 ppid=25 worker=0
2018-10-05 23:42:56 +0000 [info]: #0 Cloudfront verbose logging enabled
2018-10-05 23:42:56 +0000 [warn]: #0 S3 client error. uninitialized constant Aws::S3
2018-10-05 23:42:56 +0000 [info]: #0 fluentd worker is now running worker=0
```

# Gem list
```
*** LOCAL GEMS ***

aws-eventstream (1.0.1)
aws-partitions (1.105.0)
aws-sdk-core (3.30.0)
aws-sdk-kms (1.9.0)
aws-sdk-s3 (1.21.0)
aws-sdk-sqs (1.7.0)
aws-sigv4 (1.0.3)
bigdecimal (1.2.8)
cool.io (1.5.3)
did_you_mean (1.0.0)
dig_rb (1.0.1)
elasticsearch (6.1.0)
elasticsearch-api (6.1.0)
elasticsearch-transport (6.1.0)
excon (0.62.0)
faraday (0.15.3)
fluent-plugin-cloudfront-log-v0.14-fix (0.1.0)
fluent-plugin-elasticsearch (2.11.10)
fluentd (1.2.5)
http_parser.rb (0.6.0)
io-console (0.4.5)
jmespath (1.4.0)
json (2.1.0, 1.8.3)
minitest (5.9.0)
msgpack (1.2.4)
multi_json (1.13.1)
multipart-post (2.0.0)
net-telnet (0.1.1)
oj (3.3.10)
power_assert (0.2.7)
psych (2.1.0)
rake (10.5.0)
rdoc (4.2.1)
serverengine (2.0.7)
sigdump (0.2.4)
strptime (0.2.3)
test-unit (3.1.7)
thread_safe (0.3.6)
tzinfo (1.2.5)
tzinfo-data (1.2018.5)
yajl-ruby (1.4.1)
```

# Config
```
<source>
  @type http
  bind 127.0.0.1
  port 8888
</source>

<source>
  @type       cloudfront_log
  log_bucket  cloudfront-logs
  log_prefix  incoming
  region      us-east-1
  tag         cloudfront
  verbose     true
</source>

<filter cloudfront>
  @type record_transformer
  <record>
    #timestamp ${record["date"]}T${record["time"]}Z
    @type cloudfront_log
    remove_keys date,time
  </record>
</filter>

<match cloudfront>
  @type elasticsearch
  host host.example.com
  scheme https
  port port
  user user
  password password

  logstash_format true
  logstash_prefix cloudfront
  time_key timestamp
  time_key_format %Y-%m-%dT%H:%M:%SZ
  utc_index true

  type_name cloudfront_log
  reconnect_on_error true
  include_tag_key true
  tag_key fluentd_tag

  buffer_type file
  buffer_path /fluentd/log/buffer/cloudfront
  buffer_chunk_limit 4M
  buffer_queue_limit 20
  buffer_queue_full_action block
  flush_interval 60
  retry_limit 17
  retry_wait 2.0
  num_threads 16
  slow_flush_log_threshold 10
</match>
```

Thank you!
Lance

**Comment by lenfree** [comment]
Hi @lancelakey , seems like it is related to ` aws-sdk-core (3.30.0)`? 
  Could you share your Dockerfile or maybe you have a Docker image that I could use to debug? 😄

**Comment by lancelakey** [comment]
I wondered similarly.
Perhaps something has changed in the aws-sdk and that's causing a problem/error?

Here's my Dockerfile!
LMK if you need anything.

### References
https://github.com/fluent/fluentd-docker-image
https://github.com/fluent/fluentd-docker-image/tree/master/v1.2/debian-onbuild
https://github.com/fluent/fluentd-docker-image/blob/master/v1.2/debian-onbuild/Dockerfile

### Dockerfile
```
#
# Fluentd log shipper - https://hub.docker.com/r/fluent/fluentd/
#

FROM fluent/fluentd:stable-debian-onbuild

RUN buildDeps="sudo make gcc g++ libc-dev ruby-dev" \
 && apt-get update \
 && apt-get install -y --no-install-recommends $buildDeps \
 && sudo gem install \
        fluent-plugin-elasticsearch \
        fluent-plugin-cloudfront-log-v0.14-fix \
 && sudo gem sources --clear-all \
 && SUDO_FORCE_REMOVE=yes \
    apt-get purge -y --auto-remove \
                  -o APT::AutoRemove::RecommendsImportant=false \
                  $buildDeps \
 && rm -rf /var/lib/apt/lists/* \
           /home/fluent/.gem/ruby/2.3.0/cache/*.gem

RUN mkdir -p /fluentd/buffer
```

**Comment by lenfree** [comment]
I added some bits to your Dockerfile to confirm whether it is related to `aws-sdk-core` version and able to make it work with specific version of gem as shown below:

```Dockerfile
FROM fluent/fluentd:stable-debian-onbuild

RUN buildDeps="sudo make gcc g++ libc-dev ruby-dev" \
 && apt-get update \
 && apt-get install -y --no-install-recommends $buildDeps \
 && sudo gem install \
        fluent-plugin-elasticsearch \
        fluent-plugin-cloudfront-log-v0.14-fix \
 && sudo gem sources --clear-all \
 && SUDO_FORCE_REMOVE=yes \
    apt-get purge -y --auto-remove \
                  -o APT::AutoRemove::RecommendsImportant=false \
                  $buildDeps \
 && rm -rf /var/lib/apt/lists/* \
           /home/fluent/.gem/ruby/2.3.0/cache/*.gem

RUN gem uninstall --force aws-sdk-core aws-sdk-kms aws-sdk-s3 aws-sdk- 
  sqs aws-sigv4 \
  aws-eventstream  \
  aws-partitions  \
  jmespath  && \
        gem install aws-sdk-core -v '2.11.21' && \
        gem install aws-sdk-s3 -v '1.8.0' && \
        gem uninstall aws-sdk-core --force -v '3.30.0' 
RUN mkdir -p /fluentd/buffer
```

and logs 
```
2018-10-07 01:48:54 +0000 [info]: parsing config file is succeeded path="fluentd/etc/fluent.conf"
2018-10-07 01:48:55 +0000 [info]: @log_bucket: cloudfront-logs
2018-10-07 01:48:55 +0000 [info]: @moved_log_bucket: cloudfront-logs
2018-10-07 01:48:55 +0000 [info]: @log_prefix: incoming
2018-10-07 01:48:55 +0000 [info]: @moved_log_prefix: _moved
2018-10-07 01:48:55 +0000 [info]: @thread_num: 4
2018-10-07 01:48:55 +0000 [info]: using configuration file: <ROOT>
  <source>
    @type http
    bind "127.0.0.1"
    port 8888
  </source>
  <source>
    @type cloudfront_log
    log_bucket "cloudfront-logs"
    log_prefix "incoming"
    region "us-east-1"
    tag "cloudfront"
    verbose "true"
  </source>
  <filter cloudfront>
    @type record_transformer
    <record>
      @type cloudfront_log
      remove_keys date,time
    </record>
  </filter>
</ROOT>
2018-10-07 01:48:55 +0000 [info]: starting fluentd-1.2.5 pid=20 ruby="2.3.3"
2018-10-07 01:48:55 +0000 [info]: spawn command to main:  cm"-Eascii-8bit:ascii-8bit", "/usr/local/bin/fluentd", "-c", "fluentd/etc/fluent.conf", "--under-supervisor"]
2018-10-07 01:48:55 +0000 [info]: gem 'fluent-plugin-cloudfront-log-v0.14-fix' version '0.1.0'
2018-10-07 01:48:55 +0000 [info]: gem 'fluent-plugin-elasticsearch' version '2.11.10'
2018-10-07 01:48:55 +0000 [info]: gem 'fluentd' version '1.2.5'
2018-10-07 01:48:55 +0000 [info]: adding filter pattern="cloudfront" type="record_transformer"
2018-10-07 01:48:55 +0000 [info]: adding source type="http"
2018-10-07 01:48:55 +0000 [info]: adding source type="cloudfront_log"
2018-10-07 01:48:55 +0000 [info]: #0 @log_bucket: cloudfront-logs
2018-10-07 01:48:55 +0000 [info]: #0 @moved_log_bucket: cloudfront-logs
2018-10-07 01:48:55 +0000 [info]: #0 @log_prefix: incoming
2018-10-07 01:48:55 +0000 [info]: #0 @moved_log_prefix: _moved
2018-10-07 01:48:55 +0000 [info]: #0 @thread_num: 4
2018-10-07 01:48:55 +0000 [info]: #0 starting fluentd worker pid=24 ppid=20 worker=0
2018-10-07 01:48:55 +0000 [info]: #0 Cloudfront verbose logging enabled
2018-10-07 01:48:58 +0000 [info]: #0 fluentd worker is now running worker=0
2018-10-07 01:53:58 +0000 [info]: #0 CloudFront Begining input going to list S#
2018-10-07 01:54:01 +0000 [warn]: #0 S3 GET list error. unable to sign request without credentials set
2018-10-07 01:58:58 +0000 [info]: #0 CloudFront Begining input going to list S3
2018-10-07 01:59:01 +0000 [warn]: #0 S3 GET list error. unable to sign request without credentials set
2018-10-07 02:03:58 +0000 [info]: #0 CloudFront Begining input going to list S3
2018-10-07 02:04:01 +0000 [warn]: #0 S3 GET list error. unable to sign request without credentials set
2018-10-07 02:08:58 +0000 [info]: #0 CloudFront Begining input going to list S3
2018-10-07 02:09:01 +0000 [warn]: #0 S3 GET list error. unable to sign request without credentials set
```

Could you confirm this on your end? 

Cheers,
Lenfree

**Comment by lancelakey** [comment]
LGTM Lenfree!

Thanks so much for your help!
